### PR TITLE
FEATURE: Graceful shutdown and process groups for child processes

### LINKED ISSUE #14: Graceful shutdown
Scenario: Running jobs in prunner and the process is shut down or restarted
Implementation: Handle interrupt signal and cancel running jobs and wait until complete (would make sense with #11) until some wait time (configurable in pipeline definition) before doing a hard stop of running tasks.

### PR DESCRIPTION
**Pull request by hlubek**
Implements shutdown via SIGINT (graceful) and SIGTERM (cancelling jobs).
Also fix an issue with child processes from scripts by setting setpgid
to start processes (which is needed to prevent signals being forwarded).

Closes #14

### CODE REVIEW

File: prunner.go [hunk 1]
```diff
@@ -76,6 +82,7 @@ func NewPipelineRunner(ctx context.Context, defs *definition.PipelinesDef, creat
 			for {
 				select {
 				case <-ctx.Done():
+					log.Debugf("PipelineRunner: context done, stopping persist loop")
```
Comment (hlubek) [inline prunner.go:85]: Can be removed

File: prunner.go [hunk 2]
```diff
@@ -175,11 +182,16 @@ var errQueueFull = errors.New("concurrency exceeded and queue limit reached for
...
 var errJobAlreadyCompleted = errors.New("job is already completed")
 var errJobNotStarted = errors.New("job is not started")
+var ErrShuttingDown = errors.New("runner is shutting down")
 
 func (r *PipelineRunner) ScheduleAsync(pipeline string, opts ScheduleOpts) (*PipelineJob, error) {
 	r.mx.Lock()
 	defer r.mx.Unlock()
 
+	if r.isShuttingDown {
+		return nil, ErrShuttingDown
```
Comment (hlubek) [inline prunner.go:192]: Make sure this is handled in server with a fitting status code

File: taskctl/executor.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+	if job.Dir == "" {
+		job.Dir = e.dir
+	}
+
+	jobID := job.Vars.Get(JobIDVariableName).(string)
+
+	log.
+		WithField("component", "executor").
+		WithField("jobID", jobID).
+		Debugf("Executing \"%s\"", command)
```
Comment (hlubek) [inline taskctl/executor.go:159]: Better use `%q`

File: taskctl/runner.go [hunk 4]
```diff
@@ -214,7 +221,9 @@ func (r *TaskRunner) Cancel() {
 	r.cancelMutex.Lock()
 	if !r.canceling {
 		r.canceling = true
-		defer logrus.Debug("runner has been cancelled")
+		defer log.
```
Comment (hlubek) [inline taskctl/runner.go:224]: Seeing it in practice this log line is not needed

File: taskctl/executor_test.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+	"io/ioutil"
+	"testing"
+	"time"
+
+	"github.com/taskctl/taskctl/pkg/executor"
+
+	"github.com/Flowpack/prunner/taskctl"
+)
+
+func TestDefaultExecutor_Execute(t *testing.T) {
```
Comment (hlubek) [inline taskctl/executor_test.go:15]: _PgidExecutor_

**Review by skurfuerst — COMMENTED** [review]
really cool, just found a little nitpick / question :)

File: prunner.go [hunk 6]
```diff
@@ -774,6 +790,77 @@ func (r *PipelineRunner) SaveToStore() {
...
 
+func (r *PipelineRunner) Shutdown(ctx context.Context) error {
+	defer func() {
+		log.
+			WithField("component", "runner").
+			Debugf("Shutting down, waiting for pending operations...")
+		r.wg.Wait()
+		// Do a final save to include recently completed jobs
+		r.SaveToStore()
+		r.wg.Wait()
```
[the remark this answers is not in the document]
Comment (hlubek) [reply prunner.go:801]: Right, `SaveToStore` is not async and we can just call it without waiting again. The previous wait is to wait for all `scheduler` calls to be finished (to have correct job completions).

I'll adjust the code and add a more descriptive comment ;)

[the remark this answers is not in the document]
Comment (skurfuerst) [reply prunner.go]: perfect, thanks :)

**Pull request merged**